### PR TITLE
Add python tests for Parquet DELTA_BINARY_PACKED encoder

### PR DESCRIPTION
**Pull request by etseidl**
## Description
During the review of #14100 there was a suggestion to add a test of writing using cudf and then reading the resulting file back with pyarrow. This PR adds the necessary python bindings to perform this test.

NOTE: there is currently an issue with encoding 32-bit values where the deltas exceed 32-bits. parquet-mr and arrow truncate the deltas for the INT32 physical type and allow values to overflow, whereas cudf currently uses 64-bit deltas, which avoids the overflow, but can result in requiring 33-bits when encoding. The current cudf behavior is allowed by the specification (and in fact is readable by parquet-mr), but using the extra bit is not in the Parquet spirit of least output file size. This will be addressed in follow-on work.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/page_enc.cu [hunk 1]
```diff
@@ -396,7 +396,9 @@ __device__ size_t delta_data_len(Type physical_type, cudf::type_id type_id, uint
   // need max dtype_len + 1 bytes for min_delta
   // one byte per mini block for the bitwidth
   // and block_size * dtype_len bytes for the actual encoded data
-  auto const block_size = dtype_len + 1 + delta::num_mini_blocks + vals_per_block * dtype_len;
+  auto const max_bits        = dtype_len * 8 + 1;
+  auto const bytes_per_block = max_bits * vals_per_block / 8;
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:400]: should this be rounded up?
```suggestion
  auto const bytes_per_block = cudf::util::div_rounding_up_unsafe(max_bits * vals_per_block, 8);
```

File: cpp/src/io/parquet/page_enc.cu [hunk 2]
```diff
@@ -396,7 +396,9 @@ __device__ size_t delta_data_len(Type physical_type, cudf::type_id type_id, uint
   // need max dtype_len + 1 bytes for min_delta
   // one byte per mini block for the bitwidth
   // and block_size * dtype_len bytes for the actual encoded data
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:398]: please update the comments to reflect the new math

**Review by vuule — CHANGES_REQUESTED** [review]
a rounding concern

File: cpp/src/io/parquet/page_enc.cu (hunk 1 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu:400]: Not really a concern because vals_per_block will always be a multiple of 8, but will change it.

Comment (vuule) [reply cpp/src/io/parquet/page_enc.cu:400]: I really thought I got something there :D

File: cpp/src/io/parquet/page_enc.cu [hunk 3]
```diff
@@ -393,13 +393,20 @@ __device__ size_t delta_data_len(Type physical_type, cudf::type_id type_id, uint
...
-  // need max dtype_len + 1 bytes for min_delta
+  // need max dtype_len + 1 bytes for min_delta (because we only encode 7 bits per byte)
   // one byte per mini block for the bitwidth
-  // and block_size * dtype_len bytes for the actual encoded data
-  auto const block_size = dtype_len + 1 + delta::num_mini_blocks + vals_per_block * dtype_len;
+  auto const mini_block_header_size = dtype_len + 1 + delta::num_mini_blocks;
+  // each encoded value can be at most sizeof(type) * 8 + 1 bits
+  auto const max_bits = dtype_len * 8 + 1;
+  // each data block will then be max_bits * values per block. vals_per_block is guaranteed to be
+  // divisible by 128 (via static assert on delta::block_size), but do save division anyway.
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:402]:
```suggestion
  // divisible by 128 (via static assert on delta::block_size), but do safe division anyway.
```

**Review by vuule — APPROVED** [review]

**Comment by GregoryKimball** [comment]
@galipremsagar Would you please take a look at this testing improvement?

File: python/cudf/cudf/tests/test_parquet.py [hunk 4]
```diff
@@ -1335,6 +1332,26 @@ def test_delta_binary(nrows, add_nulls, tmpdir):
...
+        use_dictionary=False,
+    )
+
+    # FIXME(ets): should probably not use more bits than the data type
+    try:
+        cdf2 = cudf.from_pandas(pd.read_parquet(cudf_fname))
+    except OSError:
+        if dtype == "int32" and nrows == 100000:
+            pytest.mark.xfail(
+                reason="arrow does not support 33-bit delta encoding"
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_parquet.py:1350]:
```suggestion
                reason="arrow does not support 32-bit delta encoding"
```
typo?

Comment (etseidl) [reply python/cudf/cudf/tests/test_parquet.py:1350]: No, that's actually the correct wording. The cudf delta encoder will use 33 bits when encoding 32-bit integers.  Some parquet readers are ok with this (cudf, parquet-mr), some are not (arrow-cpp, arrow-rs). I have a follow-on to this PR that will restrict the encoding to 32 bits to make everyone happy, then the xfail will go away.

File: python/cudf/cudf/_lib/cpp/io/parquet.pxd [hunk 5]
```diff
@@ -191,6 +199,8 @@ cdef extern from "cudf/io/parquet.hpp" namespace "cudf::io" nogil:
         void set_row_group_size_rows(size_type val) except +
         void set_max_page_size_bytes(size_t val) except +
         void set_max_page_size_rows(size_type val) except +
+        void enable_write_v2_headers(bool val) except +
```
Comment (mythrocks) [inline python/cudf/cudf/_lib/cpp/io/parquet.pxd:202]: I realize it isn't the fault of this current PR, but one does wish `enable_write_v2_headers` were named `set_write_v2_headers`.

**Review by mythrocks — APPROVED** [review]
Not a cuIO dev, but this looks good to me.

A minor nitpick regarding the naming of a function, but that's unrelated to this change.

File: python/cudf/cudf/_lib/cpp/io/parquet.pxd (hunk 5 above)
Comment (vuule) [reply python/cudf/cudf/_lib/cpp/io/parquet.pxd:202]: We use enable_ for bool options, so this should be consistent (for better or for worse, apparently).

File: python/cudf/cudf/core/dataframe.py [hunk 6]
```diff
@@ -6370,6 +6370,8 @@ def to_parquet(
         max_page_size_rows=None,
         storage_options=None,
         return_metadata=False,
+        use_dictionary=True,
```
Comment (galipremsagar) [inline python/cudf/cudf/core/dataframe.py:6373]: Can we document these two parameters here: https://github.com/rapidsai/cudf/blob/16051a718509c218010e6912d6f8e0fca7a7aa24/python/cudf/cudf/utils/ioutils.py#L222

File: python/cudf/cudf/tests/test_parquet.py (hunk 4 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py:1350]: TIL. Thanks @etseidl !

File: python/cudf/cudf/tests/test_parquet.py [hunk 7]
```diff
@@ -1335,6 +1332,26 @@ def test_delta_binary(nrows, add_nulls, tmpdir):
...
+            )
+    else:
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_parquet.py:1352]:
```suggestion
    except OSError as e:
        if dtype == "int32" and nrows == 100000:
            pytest.mark.xfail(
                reason="arrow does not support 33-bit delta encoding"
            )
        else:
            raise e
    else:
```
We will have to raise for else case.

**Review by galipremsagar — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/dataframe.py (hunk 6 above)
Comment (etseidl) [reply python/cudf/cudf/core/dataframe.py:6373]: done

Comment (etseidl) [reply python/cudf/cudf/core/dataframe.py:6373]: Thanks @galipremsagar. I would have never thought to look there for the docstring :sweat_smile:

Comment (vuule) [reply python/cudf/cudf/core/dataframe.py:6373]: First time hearing about it as well :man_shrugging: (just don't git blame `ioutils.py` :P )

**Review by galipremsagar — APPROVED** [review]

**Pull request merged**